### PR TITLE
nvmeof-recoverer: add NvmeOfStorage CRD

### PR DESCRIPTION
**Pull request by cheolho-kang**
**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by myoungwon** [comment]
@cheolho-kang I think the second and third commits can be merged into one.

**Comment by myoungwon** [comment]
Please correct the title.

File: pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go [hunk 1]
```diff
@@ -60,12 +69,53 @@ type ReconcileNvmeOfStorage struct {
...
+	err = c.Watch(cmKind, &handler.EnqueueRequestForObject{}, opcontroller.WatchControllerPredicate())
+	if err != nil {
+		return err
+	}
+
 	return nil
 }
+
+func (r *ReconcileNvmeOfStorage) Reconcile(context context.Context, request reconcile.Request) (reconcile.Result, error) {
+	reconcileResponse, err := r.reconcile(request)
```
Comment (myoungwon) [inline pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go]: I don't think this is necessary here.

Comment (myoungwon) [reply pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go]: Please return directly with a correct return value.

Comment (cheolho-kang) [reply pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go]: Yes, it seems unnecessary as you mentioned. I will merge them into one.
One thing that concerns me is that the existing `controller`s are split into `Reconcile` and `reconcile`.
Doesn't it need to match the style of the existing code? However the other controllers didn't seem like there was a particular reason for splitting them up either.

File: deploy/examples/common.yaml [hunk 2]
```diff
@@ -84,7 +84,17 @@ metadata:
     app.kubernetes.io/name: cosi-driver-ceph
 rules:
   - apiGroups: ["objectstorage.k8s.io"]
-    resources: ["buckets", "bucketaccesses", "bucketclaims", "bucketaccessclasses", "buckets/status", "bucketaccesses/status", "bucketclaims/status", "bucketaccessclasses/status"]
```
Comment (myoungwon) [inline deploy/examples/common.yaml]: Why did you modify this line?

File: deploy/examples/common.yaml [hunk 3]
```diff
@@ -581,7 +594,8 @@ rules:
     verbs: ["create"]
   - apiGroups: ["csiaddons.openshift.io"]
     resources: ["networkfences"]
-    verbs: ["create", "get", "update", "delete", "watch", "list", "deletecollection"]
```
Comment (myoungwon) [inline deploy/examples/common.yaml]: same here

File: deploy/examples/common.yaml [hunk 4]
```diff
@@ -858,7 +872,7 @@ rules:
       - patch
       - delete
   - apiGroups:
-      - ''
```
Comment (myoungwon) [inline deploy/examples/common.yaml]: same here

File: deploy/examples/common.yaml (hunk 2 above)
Comment (cheolho-kang) [reply deploy/examples/common.yaml]: This was changed when it was auto-generated, but I'm not sure why it was generated differently..
I will fix it with the original code

File: deploy/examples/common.yaml (hunk 3 above)
Comment (cheolho-kang) [reply deploy/examples/common.yaml]: Okay, I will fix it

File: deploy/examples/common.yaml (hunk 4 above)
Comment (cheolho-kang) [reply deploy/examples/common.yaml]: Okay, I will fix it

File: pkg/client/informers/externalversions/factory.go [hunk 5]
```diff
@@ -142,7 +142,7 @@ func (f *sharedInformerFactory) WaitForCacheSync(stopCh <-chan struct{}) map[ref
 	return res
 }
 
-// InformerFor returns the SharedIndexInformer for obj using an internal
```
Comment (myoungwon) [inline pkg/client/informers/externalversions/factory.go]: same here

Comment (cheolho-kang) [reply pkg/client/informers/externalversions/factory.go]: Fix it

**Pull request merged**